server/src/main/java/io/crate/planner/optimizer/rule/RewriteFilterOnOuterJoinToInnerJoin.java

```diff
@@ -272,4 +275,34 @@ private static boolean couldMatchOnNull(@Nullable Symbol query,
         Input<?> input = query.accept(evaluator, null);
         return WhereClause.canMatch(input);
     }
+
+    private static Symbol[] divideIntoLeftAndRightQueries(Map<Set<RelationName>, Symbol> splitQueries,
+                                                          Set<RelationName> lhsRelations,
+                                                          Set<RelationName> rhsRelations) {
+        Symbol leftQuery = splitQueries.remove(lhsRelations);
+        Symbol rightQuery = splitQueries.remove(rhsRelations);
+
+        if (leftQuery == null && rightQuery == null) {
+            // we don't have a match for the filter on rhs/lhs yet
+            // let's see if we have partial match with a subsection of the relations
+            var it = splitQueries.entrySet().iterator();
+            while (it.hasNext()) {
+                var entry = it.next();
+                var relationNames = entry.getKey();
+                var query = entry.getValue();
+
+                var matchesLhs = Sets.intersection(lhsRelations, relationNames);
+                var matchesRhs = Sets.intersection(rhsRelations, relationNames);
+
+                if (matchesRhs.isEmpty() == false && matchesLhs.isEmpty()) {
+                    rightQuery = query;
+                    it.remove();
+                } else if (matchesRhs.isEmpty() && matchesLhs.isEmpty() == false) {
+                    leftQuery = query;
+                    it.remove();
+                }
```

Comment:
This is copied from `MoveFilterBeneathJoin`. At this point, I feel `MoveFilterBeneathJoin` and `RewriteFilterOnOuterJoinToInnerJoin` are overlapping.

Reply:
Will look into this after this PR is merged.